File: Ruby短信接口/IKYunSender.rb
require 'time'
require 'digest'
require 'net/https'
class IKYunSender#No indent python example writen by ouyang2008
    def initialize(apiurl,appid,appkey)
        @mapiurl=apiurl
        @mappid=appid
        @mappkey=appkey
    end
    def send(mobile,extno,msg)
        timestamp=self._getTimeStamp()
        puts(timestamp)
        digest=self._signDigest(mobile,timestamp)
        params={"ac":"sendsms","mobile":mobile,"extno":extno,"appid":@mappid,"digest":digest,"msg":msg,"timestamp":timestamp}
        puts(params)
        uri = URI(@mapiurl)
        ret=Net::HTTP.post_form(uri,params)
        return ret.body
    end
    def getAmount()
        timestamp=self._getTimeStamp()
        puts(timestamp)
        digest=self._signDigest("",timestamp)
        params={"ac":"getamount","appid":@mappid,"digest":digest,"timestamp":timestamp}
        puts(params)
        uri = URI(@mapiurl)
        ret=Net::HTTP.post_form(uri,params)
        return ret.body
    end
    def getReplyMsg()
        timestamp=self._getTimeStamp()
        puts(timestamp)
        digest=self._signDigest("",timestamp)
        params={"ac":"fetchreply","appid":@mappid,"digest":digest,"timestamp":timestamp}
        uri = URI(@mapiurl)
        ret=Net::HTTP.post_form(uri,params)
        return ret.body
    end
    def _signDigest(mobile,timestamp)
        m1=Digest::MD5.new
        m1.update(@mappkey.encode("utf-8"))
        digestappkey=m1.hexdigest
        puts(digestappkey)
        srcd=@mappid+digestappkey+timestamp+mobile
        puts(srcd)
        m2=Digest::MD5.new
        m2.update(srcd.encode("utf-8"))
        return m2.hexdigest
    end
    def _getTimeStamp()
        return Time.now.utc.to_i.to_s
    end
    def test
    puts "rb test !"
    end
end

File: python短信接口/python/ikyunsender.py
import http.client
import hashlib
import time
import urllib
import urllib.parse
class IkYunSender:#code by ouyang2008,and it is free
    def __init__(self,apiurl,appid,appkey):
        self.ApiUrl=apiurl
        self.AppID=appid
        self.AppKey=appkey
        self._parseApiUrl()
    def send(self,mobile,extno,msg):
        timestamp=self._getTimeStamp()
        digest=self._signdigest(mobile,timestamp)
        params = {"ac":'sendsms','mobile': mobile,'appid':self.AppID, 'extno': extno,'digest':digest,'timestamp':timestamp,'msg':msg}
        print(params)
        return self._post(urllib.parse.urlencode(params))

    def _signdigest(self,mobile,timestamp):
         m = hashlib.md5()
         m.update(self.AppKey.encode("utf-8"))
         appkeydigest=m.hexdigest()
         print(appkeydigest)
         #print appkedigest
         m2=hashlib.md5()
         m2.update((self.AppID+appkeydigest+timestamp+mobile).encode("utf-8"))
         hashstr=m2.hexdigest()
         return hashstr

    def _post(self,data):
        conn = None
        try:
            headers = {"Content-type": "application/x-www-form-urlencoded;charset=utf8"}
            print(self.Host)
            conn = http.client.HTTPConnection(self.Host, 80, timeout=200)
            print(self.Uri)
            conn.request("POST", self.Uri, data, headers)
            response = conn.getresponse()
            retmsg = response.read().decode("utf8")
            conn.close()
            return retmsg
        except  e:
           print(e)
        finally:
           if conn:
              conn.close()

    def getAmount(self):
        timestamp=self._getTimeStamp()
        digest=self._signdigest('',timestamp)
        params = {"ac":'getamount','appid':self.AppID, 'digest':digest,'timestamp':timestamp}
        print(params)
        return self._post(urllib.parse.urlencode(params))
    def getReplyMsg(self):
        timestamp=self._getTimeStamp()
        digest=self._signdigest('',timestamp)
        params = {"ac":'fetchreply','appid':self.AppID, 'digest':digest,'timestamp':timestamp}
        print(params)
        return self._post(urllib.parse.urlencode(params))
    def _getTimeStamp(self):
        return str(int(time.time()))
    def _parseApiUrl(self):
        urltup=urllib.parse.urlparse(self.ApiUrl)
        self.Host=urltup.hostname
        self.Uri=urltup.path
